### PR TITLE
[SPARK-28763][SQL][TEST] Flaky Tests: SparkThriftServerProtocolVersionsSuite.HIVE_CLI_SERVICE_PROTOCOL_V1 get binary type

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?
This pr try to fix flaky tests:
```
org.scalatest.exceptions.TestFailedException: "[?](" did not equal "[�]("
```

```
org.apache.spark.sql.hive.thriftserver.SparkThriftServerProtocolVersionsSuite.HIVE_CLI_SERVICE_PROTOCOL_V1 get binary type | 0.48 秒 | 2
-- | -- | --
org.apache.spark.sql.hive.thriftserver.SparkThriftServerProtocolVersionsSuite.HIVE_CLI_SERVICE_PROTOCOL_V2 get binary type | 0.34 秒 | 2
org.apache.spark.sql.hive.thriftserver.SparkThriftServerProtocolVersionsSuite.HIVE_CLI_SERVICE_PROTOCOL_V3 get binary type | 0.36 秒 | 2
org.apache.spark.sql.hive.thriftserver.SparkThriftServerProtocolVersionsSuite.HIVE_CLI_SERVICE_PROTOCOL_V4 get binary type | 0.3 秒 | 2
org.apache.spark.sql.hive.thriftserver.SparkThriftServerProtocolVersionsSuite.HIVE_CLI_SERVICE_PROTOCOL_V5 get binary type | 0.36 秒 | 2
```

https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4832/testReport/

### How was this patch tested?
N/A

### CODE REVIEW

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala [hunk 1]
```diff
@@ -222,7 +222,8 @@ class SparkThriftServerProtocolVersionsSuite extends HiveThriftJdbcTest {
         assert(rs.next())
         assert(rs.getString(1) === "ABC")
       }
-      testExecuteStatementWithProtocolVersion(version, "SELECT cast(49960 as binary)") { rs =>
+      testExecuteStatementWithProtocolVersion(version,
+        "SELECT cast(cast(49960 as int) as binary)") { rs =>
```
Comment (HyukjinKwon) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: @wangyum Btw how does this fix the flakiness?

**Comment by dongjoon-hyun** [comment]
Thank you for the fix. I also have the same question like @HyukjinKwon .
Could you describe more about the root cause in the PR description?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala [hunk 2]
```diff
@@ -224,7 +224,9 @@ class SparkThriftServerProtocolVersionsSuite extends HiveThriftJdbcTest {
       }
       testExecuteStatementWithProtocolVersion(version, "SELECT cast(49960 as binary)") { rs =>
         assert(rs.next())
-        assert(rs.getString(1) === UTF8String.fromBytes(NumberConverter.toBinary(49960)).toString)
+        assertResult(Array(0, 0, 65533, 40).map(_.toChar)) {
+          rs.getString(1).toCharArray
+        }
```
Comment (wangyum) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: @dongjoon-hyun @HyukjinKwon Maybe we should uses CharArray.

Comment (dongjoon-hyun) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: What is the difference? Could you tell us the root cause you think?

Comment (wangyum) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: I'm not sure. May be related to environment:
https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4832/
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109194/testReport/

These two PullRequestBuilders are built on the same machine(amp-jenkins-worker-03). The first one failed, but the last one successful. The environment differences:
image

Comment (wangyum) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: I will verify it later.

Comment (dongjoon-hyun) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: Maybe, something like `LANG=en_US.UTF_8`?

**Comment by dongjoon-hyun** [comment]
@wangyum . There are 8 protocols and `V6/V7/V8` passed with `assert(rs.getString(1) === UTF8String.fromBytes(NumberConverter.toBinary(49960)).toString)`. For all 8 protocol, the assertion shares the same environment for `V1~V8`. So, the assertion looks correct to me.
```
[info] - HIVE_CLI_SERVICE_PROTOCOL_V1 get binary type *** FAILED *** (486 milliseconds)
[info] - HIVE_CLI_SERVICE_PROTOCOL_V2 get binary type *** FAILED *** (345 milliseconds)
[info] - HIVE_CLI_SERVICE_PROTOCOL_V3 get binary type *** FAILED *** (361 milliseconds)
[info] - HIVE_CLI_SERVICE_PROTOCOL_V4 get binary type *** FAILED *** (304 milliseconds)
[info] - HIVE_CLI_SERVICE_PROTOCOL_V5 get binary type *** FAILED *** (360 milliseconds)
[info] - HIVE_CLI_SERVICE_PROTOCOL_V6 get binary type (521 milliseconds)
[info] - HIVE_CLI_SERVICE_PROTOCOL_V7 get binary type (588 milliseconds)
[info] - HIVE_CLI_SERVICE_PROTOCOL_V8 get binary type (565 milliseconds)
```

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala (hunk 2 above)
Comment (wangyum) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: Yes. I can reproduce it by `unset LANG`:
```
[info] - HIVE_CLI_SERVICE_PROTOCOL_V1 get binary type *** FAILED *** (458 milliseconds)
[info]   "[?](" did not equal "[?](" (SparkThriftServerProtocolVersionsSuite.scala:227)
```

Comment (srowen) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: @shaneknapp I think we need to standardize `LANG` across the workers? or should I try giving it shot?
That seems reasonable though I am distantly concerned that something in the code relies on this setting. That said, I think it's probably correct to set `LANG` in a prod environment and correct to set it to this value anyway.

Comment (shaneknapp) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: > I'm not sure. May be related to environment:
> https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4832/
> https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109194/testReport/
> 
> These two PullRequestBuilders are built on the same machine(amp-jenkins-worker-03). The first one failed, but the last one successful. The environment differences:
> image

ok, the SparkPullRequestBuilder is triggered by the amplab jenkins bot, and the NewSparkPullRequestBuilder is triggered by databricks' test site (http://spark-prs.appspot.com).

i can easily add the LANG variable in the former build.  another option is to add it to the default environment vars for all workers.

Comment (shaneknapp) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: (fyi, i just added LANG=en_US.UTF-8 to the NewSparkPullRequestBuilder job)

Comment (srowen) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: @shaneknapp I'd set it everywhere, personally, if it's easy.

Comment (shaneknapp) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: it's easy, but requires each worker to be disconnected/reconnected...  i'll get that set up now but it won't go in to effect until i have the opportunity to restart things.

Comment (wangyum) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: Thank you @shaneknapp

Comment (wangyum) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: Could we close this PR?

Comment (dongjoon-hyun) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkThriftServerProtocolVersionsSuite.scala]: Thank you for investigating this so far, @wangyum . Yes. It looks like that. 
You can close this and the JIRA together. If the failure is reported again, we can reopen this.

**Pull request closed without merging**